lib/ssz_ex.ex

```diff
@@ -184,6 +192,145 @@ defmodule LambdaEthereumConsensus.SszEx do
     end
   end
 
+  defp encode_container(container, schemas) do
+    {fixed_size_values, fixed_length, variable_values} = analyze_schemas(container, schemas)
+
+    with {:ok, variable_parts} <- encode_schemas(variable_values),
+         offsets = calculate_offsets(variable_parts, fixed_length),
+         variable_length =
+           Enum.reduce(variable_parts, 0, fn part, acc -> byte_size(part) + acc end),
+         :ok <- check_length(fixed_length, variable_length),
+         {:ok, fixed_parts} <-
+           replace_offsets(fixed_size_values, offsets)
+           |> encode_schemas do
+      (fixed_parts ++ variable_parts)
+      |> Enum.join()
+      |> then(&{:ok, &1})
+    end
+  end
+
+  defp analyze_schemas(container, schemas) do
+    schemas
+    |> Enum.reduce({[], 0, []}, fn {key, schema},
+                                   {acc_fixed_size_values, acc_fixed_length, acc_variable_values} ->
+      value = Map.fetch!(container, key)
+
+      if variable_size?(schema) do
+        {[:offset | acc_fixed_size_values], @bytes_per_length_offset + acc_fixed_length,
+         [{value, schema} | acc_variable_values]}
+      else
+        {[{value, schema} | acc_fixed_size_values], acc_fixed_length + get_fixed_size(schema),
+         acc_variable_values}
+      end
+    end)
+  end
+
+  defp encode_schemas(tuple_values) do
+    Enum.map(tuple_values, fn {value, schema} -> encode(value, schema) end)
+    |> flatten_results()
+  end
+
+  defp calculate_offsets(variable_parts, fixed_length) do
+    {offsets, _} =
+      Enum.reduce(variable_parts, {[], 0}, fn element, {res, acc} ->
+        sum = fixed_length + acc
+        {[{sum, {:int, 32}} | res], byte_size(element) + acc}
+      end)
+
+    offsets
+  end
+
+  defp replace_offsets(fixed_size_values, offsets) do
+    {fixed_size_values, _} =
+      Enum.reduce(fixed_size_values, {[], offsets}, fn element,
+                                                       {acc_fixed_list, acc_offsets_list} ->
+        if element == :offset do
+          [offset | rest_offsets] = acc_offsets_list
+          {[offset | acc_fixed_list], rest_offsets}
+        else
+          {[element | acc_fixed_list], acc_offsets_list}
+        end
+      end)
+
+    fixed_size_values
+  end
+
+  defp decode_container(binary, module) do
```

Comment:
I like the implementation! Specially that this function is clearly divided in three. The `build_deserialization_strategy` though does a bit too many things and misses some opportunities to perform some other work. I suggest refactoring this a bit into something like this:

```elixir
fixed_length = get_fixed_length(schemas)
<<fixed_binary::bitstring-size(fixed_length), variable_binary::bitstring>> = binary
with {:ok, fixed_parts, offsets} <- decode_fixed_section(fixed_binary, schemas),
     {:ok, variable_parts} <- decode_variable_section(variable_binary, offsets) do
       {:ok, struct!(fixed_parts ++ variable_parts)
end
```

Some clarifications:

```elixir
# Returns {:ok, fixed_parts, offsets, variable_binary} unless there are errors.
# fixed_parts is a list of {key, decoded_value} pairs.
# offsets is a list of {offset, schema} pairs. The first offset is 0.
defp decode_fixed_section(binary, schemas)
```

This function should be a reduce over the schemas, similar to the one you coded, but it also decodes, and the offset list is built differently (the offsets are mapped to schemas and have the fixed_length subtracted).

```elixir
# Returns {:ok, [{key, decoded_value}]}.
defp decode_variable_section
```

This function should reduce over the offsets, and in each step we get a chunk of bytes from the variable section and use the schema to decode them.

Reply:
:thinking: I am not sure to understand the `decode_variable_section` function. I did almost the same approach of creating overlapping subslices over the offsets list. Since the variable elements do not have a fixed length, I think, it is necessary to check the next offset in order to know the length of the variable binary. Let me know if there is another approach that I am not seeing.

Reply:
I think your approach is great and follows the idea I was getting at!